client/client.go

```diff
@@ -77,151 +75,125 @@ func NewClientFromFile(certFile, keyFile, caFile string) (*Client, error) {
+	for _, host := range cert.DNSNames {
+		if ips, err := net.LookupIP(host); err == nil {
+			for _, ip := range ips {
+				c.Blacklist.Add(&net.TCPAddr{
+					IP:   ip,
+					Port: port,
+				})
 			}
-			c.m.Unlock()
 		}
 	}
+}
 
-	// choose from possible servers at random until a connection can be established.
-	for len(servers) > 0 {
-		n := rand.Intn(len(servers))
-		conn, err := c.Dial(servers[n])
-		if err == nil {
-			return conn, nil
-		}
-		log.Debugf("Couldn't dial server %s: %v", servers[n], err)
-		servers = append(servers[:n], servers[n+1:]...)
-	}
-	return nil, errors.New("couldn't dial any of the registered servers")
+// ClearBlacklist empties the client blacklist
+func (c *Client) ClearBlacklist() {
+	c.Blacklist = make(AddrSet)
 }
 
-// getServers returns the keyserver that have been registered with the given SKI.
-func (c *Client) getServers(ski gokeyless.SKI) []string {
+// Dial smartly establishes a connection to a registered keyless server
+// or reuses an existing connection if possible.
+func (c *Client) Dial(ski gokeyless.SKI) (*gokeyless.Conn, error) {
 	c.m.RLock()
-	defer c.m.RUnlock()
-	return c.allServers[ski]
+	r, ok := c.remotes[ski]
+	c.m.RUnlock()
+	if !ok {
+		if c.DefaultRemote != nil {
+			return nil, fmt.Errorf("no servers registered for SKI %02x", ski)
+		}
+
+		r = c.DefaultRemote
+	}
+	return r.Dial(c)
 }
 
 // ActivateServer dials a server and sends an activation request.
 func (c *Client) ActivateServer(server string, token []byte) error {
-	conn, err := c.Dial(server)
+	r, err := c.LookupServer(server)
 	if err != nil {
 		return err
 	}
-	defer conn.Close()
-
-	return conn.Activate(token)
-}
 
-// PopulateBlacklist populates the client blacklist using an x509 certificate.
-func (c *Client) PopulateBlacklist(cert *x509.Certificate, port string) {
-	c.BlacklistPort = port
-	c.BlacklistIPs = append(c.BlacklistIPs, cert.IPAddresses...)
-	for _, host := range cert.DNSNames {
-		if ips, err := net.LookupIP(host); err == nil {
-			c.BlacklistIPs = append(c.BlacklistIPs, ips...)
-		}
+	conn, err := r.Dial(c)
+	if err != nil {
+		return err
 	}
-}
+	defer conn.Close()
 
-// ClearBlacklist empties the client blacklist
-func (c *Client) ClearBlacklist() {
-	c.BlacklistPort = ""
-	c.BlacklistIPs = c.BlacklistIPs[:0]
+	return conn.Activate(token)
 }
 
-// ValidServer determines if the given server can be resolved and doesn't resolve
-// to a blacklisted server IP.
-func (c *Client) ValidServer(server string) (err error) {
-	var host, port string
-	if host, port, err = net.SplitHostPort(server); err != nil {
-		return
-	}
-	if port != c.BlacklistPort {
-		return nil
-	}
+// AddRemote adds a remote for the given SKI, optionally registering it
+// under the given server name.
+func (c *Client) AddRemote(server string, r Remote, ski gokeyless.SKI) *PrivateKey {
+	c.m.Lock()
+	defer c.m.Unlock()
 
-	var ips []net.IP
-	if ips, err = net.LookupIP(host); err != nil {
-		return
+	if server != "" {
```

Comment:
I could use some eyes on all the cached `c.servers` and `c.remotes` logic to make sure it make sense.